tests/models/test_taskinstance.py

```diff
@@ -992,3 +994,116 @@ def success_handler(self, context):
+                                   catchup: bool) -> dict:
+        dag_id = 'test_previous_dates'
+        dag = models.DAG(dag_id=dag_id, schedule_interval=schedule_interval, catchup=catchup)
+        task = DummyOperator(task_id='task', dag=dag, start_date=DEFAULT_DATE)
+
+        def get_test_ti(session, execution_date: str, state: str) -> TI:
+            dag.create_dagrun(
+                run_id='scheduled__{}'.format(execution_date),
+                state=state,
+                execution_date=pendulum.parse(execution_date),
+                start_date=pendulum.utcnow(),
+                session=session
+            )
+            ti = TI(task=task, execution_date=pendulum.parse(execution_date))
+            ti.set_state(state=State.SUCCESS, session=session)
+            return ti
+
+        with create_session() as session:  # type: Session
+
+            d1_s = '2019-01-01T00:00:00+00:00'
+            t1_s = get_test_ti(session, d1_s, State.SUCCESS)
+
+            d2_f = '2019-01-02T00:00:00+00:00'
+            t2_f = get_test_ti(session, d2_f, State.FAILED)
+
+            d3_s = '2019-01-03T00:00:00+00:00'
+            t3_s = get_test_ti(session, d3_s, State.SUCCESS)
+
+            return {
+                'ed1_s': pendulum.parse(d1_s),
+                'ed2_f': pendulum.parse(d2_f),
+                'ed3_s': pendulum.parse(d3_s),
+                'ti1_s': t1_s,
+                'ti2_f': t2_f,
+                'ti3_s': t3_s,
+            }
+
+    @parameterized.expand((
+        ('cron/catchup', '0 0 * * * ', True),
+        ('cron/no-catchup', '0 0 * * *', False),
+        ('no-sched/catchup', None, True),
+        ('no-sched/no-catchup', None, False),
+        ('timedelta/catchup', datetime.timedelta(days=1), True),
+        ('timedelta/no-catchup', datetime.timedelta(days=1), False),
+    ))
+    def test_previous_ti(self, _, schedule_interval, catchup) -> None:
+
+        setup_dict = self._test_previous_dates_setup(schedule_interval, catchup)
+
+        self.assertIsNone(setup_dict['ti1_s'].previous_ti)
+
+        self.assertEqual(
+            setup_dict['ti3_s'].previous_ti.execution_date,
+            setup_dict['ed2_f']
+        )
+
+    @parameterized.expand((
+        ('cron/catchup', '0 0 * * * ', True),
+        ('cron/no-catchup', '0 0 * * *', False),
+        ('no-sched/catchup', None, True),
+        ('no-sched/no-catchup', None, False),
+        ('timedelta/catchup', datetime.timedelta(days=1), True),
+        ('timedelta/no-catchup', datetime.timedelta(days=1), False),
+    ))
+    def test_previous_ti_success(self, _, schedule_interval, catchup) -> None:
+
+        setup_dict = self._test_previous_dates_setup(schedule_interval, catchup)
+
+        self.assertIsNone(setup_dict['ti1_s'].previous_ti_success)
+
+        self.assertEqual(
+            setup_dict['ti3_s'].previous_ti_success.execution_date,
+            setup_dict['ed1_s']
```

Comment:
Could we do:

```suggestion
            setup_dict['ti3_s'].execution_date
```

then we wouldn't need to return the dates directly, just the task instances? (It took me longer than I like to admit to work out what the `ed` prefix meant!)

Reply:
> It took me longer than I like to admit to work out what the ed prefix meant!
I am sorry, I can make verbose if you think better.  I should have done `ed_1_s` in hindsight.

The reason I used the actual dates is to ensure that these comparisons were not trivially equal because of some default under the hood.  E.g., I figured,  maybe every task get's a default execution_date under certain circumstances.  Then they could pass tests but they are not actually the correct dates.  

If you think this is silly I can chop the dates. LMK!

Reply:
This issue is rendered obsolete by latest update.  And to guard against trivial equality, I added some `assertNotEqual` asserts.  Marking resolved.